smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -532,20 +569,326 @@ private void generateOperationDeserializer(
+            writer.openBlock("let $L: $L = {", "};", contentName, outputType, () -> {
+                writer.write("$$metadata: deserializeMetadata(output),");
+                writer.write("__type: $S,", operation.getOutput().get().toString());
+            });
+            readHeaders(context, operation, bindingIndex);
+            readResponseBody(context, operation, bindingIndex);
+            writer.write("return Promise.resolve($L);", contentName);
         });
+        writer.write("");
+
+        // Write out the error deserialization dispatcher.
+        writer.openBlock("async function $L(\n"
+                       + "  output: $T,\n"
+                       + "  context: SerdeContext,\n"
+                       + "): Promise<$L> {", "}", errorMethodName, responseType, outputType, () -> {
+            writer.write("let data: any = await parseBody(output.body, context);");
+            writer.write("let response: any;");
+            writeErrorDeserializationDispatcher(context, operation.getErrors(), operation.getId().getNamespace());
+            writer.write("return Promise.reject(response);");
+        });
+        writer.write("");
+    }
 
+    private void readHeaders(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        String contentName = StringUtils.uncapitalize(operation.getOutput().get().getName());
+
+        ShapeIndex index = context.getModel().getShapeIndex();
+        for (HttpBinding binding : bindingIndex.getRequestBindings(operation, Location.HEADER)) {
+            String memberName = symbolProvider.toMemberName(binding.getMember());
+            writer.openBlock("if (output.headers[$L] !== undefined) {", "}", binding.getLocationName(), () -> {
+                Shape target = index.getShape(binding.getMember().getTarget()).get();
+                String headerValue = getOutputValue(context, binding.getLocation(),
+                        "output.headers[" + binding.getLocationName() + "]", operation, target);
+                writer.write("$L.$L = $L;", contentName, memberName, headerValue);
+            });
+        }
+
+        // Handle loading up prefix headers.
+        List<HttpBinding> prefixHeaderBindings = bindingIndex.getResponseBindings(operation, Location.PREFIX_HEADERS);
+        if (!prefixHeaderBindings.isEmpty()) {
+            // Run through the headers one time, matching any prefix groups.
+            writer.openBlock("Objects.keys(output.headers).forEach(header -> {", "});", () -> {
+                for (HttpBinding binding : prefixHeaderBindings) {
+                    // Generate a single block for each group of prefix headers.
+                    writer.openBlock("if (header.startsWith($L) {", binding.getLocationName(), "}", () -> {
```

Comment:
I think you're missing a ")" after `startsWith($L)`... We should really add integration tests that tries to compile the output of various models.